crates/rustc_codegen_spirv/src/codegen_cx/constant.rs

```diff
@@ -290,24 +298,7 @@ impl ConstCodegenMethods for CodegenCx<'_> {
                         (self.get_static(def_id), AddressSpace::DATA)
                     }
                 };
-                let value = if offset.bytes() == 0 {
-                    base_addr
-                } else {
-                    self.tcx
-                        .dcx()
-                        .fatal("Non-zero scalar_to_backend ptr.offset not supported")
-                    // let offset = self.constant_bit64(ptr.offset.bytes());
-                    // self.gep(base_addr, once(offset))
-                };
-                if let Primitive::Pointer(_) = layout.primitive() {
-                    assert_ty_eq!(self, value.ty, ty);
-                    value
-                } else {
-                    self.tcx
-                        .dcx()
-                        .fatal("Non-pointer-typed scalar_to_backend Scalar::Ptr not supported");
-                    // unsafe { llvm::LLVMConstPtrToInt(llval, llty) }
-                }
+                self.const_bitcast(self.const_ptr_byte_offset(base_addr, offset), ty)
```

Comment:
This drive-by change is the minimal subset of backporting this (not yet landed) upstream PR:
- https://github.com/rust-lang/rust/pull/142960

I already have later changes to this code that bring it even closer to that version, but I only did this part for the special-case of "reading a `Scalar::Ptr` as an integer", because the old `assert_ty_eq!` would fail (while compiling libcore, IIRC?) even though the whole point is to end in a `const_bitcast` regardless of what `ty` is.